### PR TITLE
added more specific http status codes

### PR DESCRIPTION
**Pull request by miiu96**

### CODE REVIEW

File: api/shared/shared.go [hunk 1]
```diff
@@ -11,7 +11,17 @@ import (
...
-func RespondWith(c *gin.Context, status int, dataField interface{}, error string, code data.ReturnCode) {
+func RespondWith(c *gin.Context, status int, dataField interface{}, error string) {
+	var code data.ReturnCode
+	switch {
+	case status >= 200 && status < 300:
+		code = data.ReturnCodeSuccess
+	case status >= 400 && status < 500:
+		code = data.ReturnCodeRequestError
+	case status >= 500 && status < 600:
+		code = data.ReturnCodeInternalError
```
Comment (bogdan-rosianu) [inline api/shared/shared.go:22]: `default`? where you will put `ReturnCodeInternalError`

File: facade/interface.go [hunk 2]
```diff
@@ -48,24 +48,24 @@ type ValidatorStatisticsProcessor interface {
 // NodeStatusProcessor defines what a node status processor should do
 type NodeStatusProcessor interface {
 	GetNetworkConfigMetrics() (*data.GenericAPIResponse, error)
-	GetNetworkStatusMetrics(shardID uint32) (*data.GenericAPIResponse, error)
+	GetNetworkStatusMetrics(shardID uint32) (*data.GenericAPIResponse, int, error)
 	GetEconomicsDataMetrics() (*data.GenericAPIResponse, error)
```
Comment (bogdan-rosianu) [inline facade/interface.go:52]: why didn't you add the status code eveywhere?

File: process/accountProcessor.go [hunk 3]
```diff
@@ -40,44 +40,52 @@ func NewAccountProcessor(proc Processor, pubKeyConverter core.PubkeyConverter, c
...
-		_, err = ap.proc.CallGetRestEndPoint(observer.Address, AddressPath+address, responseAccount)
+		httpStatus, err := ap.proc.CallGetRestEndPoint(observer.Address, AddressPath+address, responseAccount)
 		if err == nil {
 			log.Info("account request", "address", address, "shard ID", observer.ShardId, "observer", observer.Address)
-			return &responseAccount.Data.AccountData, nil
+			return &responseAccount.Data.AccountData, http.StatusOK, nil
 		}
 
 		log.Error("account request", "observer", observer.Address, "address", address, "error", err.Error())
+		gHttpStatus = httpStatus
```
Comment (bogdan-rosianu) [inline process/accountProcessor.go:75]: `g`?

File: process/accountProcessor.go [hunk 4]
```diff
@@ -40,44 +40,52 @@ func NewAccountProcessor(proc Processor, pubKeyConverter core.PubkeyConverter, c
...
 	}
 
-	return nil, ErrSendingRequest
+	return nil, gHttpStatus, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/accountProcessor.go:78]: I don't think it is correct. From what I see, you will return the latest http status.
Imagine that all observers are down. Then here, you will return `404`. When parsing the code, 400 is between [400, 500) so you will return `bad request` but it's not the caller's fault so I would rather return `InternalServerError` directly here.

File: process/accountProcessor.go [hunk 5]
```diff
@@ -89,25 +97,27 @@ func (ap *AccountProcessor) GetValueForKey(address string, key string) (string,
...
-			return apiResponse.Data.Value, nil
+			return apiResponse.Data.Value, respCode, nil
 		}
 
 		log.Error("account value for key request", "observer", observer.Address, "address", address, "error", err.Error())
+		gHttpStatus = respCode
 	}
 
-	return "", ErrSendingRequest
+	return "", gHttpStatus, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/accountProcessor.go:110]: same here. `all observers down/not responding != bad request `

File: process/accountProcessor.go [hunk 6]
```diff
@@ -120,25 +130,27 @@ func (ap *AccountProcessor) GetESDTTokenData(address string, key string) (*data.
...
-			return &apiResponse, nil
+			return &apiResponse, http.StatusOK, nil
 		}
 
 		log.Error("account get ESDT token data error", "observer", observer.Address, "address", address, "error", err.Error())
+		gHttpStatus = respCode
 	}
 
-	return nil, ErrSendingRequest
+	return nil, gHttpStatus, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/accountProcessor.go:143]: same here

File: process/accountProcessor.go [hunk 7]
```diff
@@ -150,44 +162,50 @@ func (ap *AccountProcessor) GetAllESDTTokens(address string) (*data.GenericAPIRe
...
-			return &apiResponse, nil
+			return &apiResponse, http.StatusOK, nil
 		}
 
 		log.Error("account get all ESDT tokens error", "observer", observer.Address, "address", address, "error", err.Error())
+		gHttpStatus = respCode
 	}
 
-	return nil, ErrSendingRequest
+	return nil, gHttpStatus, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/accountProcessor.go:175]: here

File: process/transactionProcessor.go [hunk 8]
```diff
@@ -134,65 +134,65 @@ func (tp *TransactionProcessor) SendTransaction(tx *data.Transaction) (int, stri
...
 			Error: response.Error,
 			Code:  response.Code,
-		}, nil
+		}, http.StatusOK, nil
 	}
 
 	observersForReceiverShard, err := tp.proc.GetObservers(receiverShardID)
 	if err != nil {
-		return nil, err
+		return nil, http.StatusInternalServerError, err
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:190]: you returned `bad request` in the same situation on other components.

File: process/transactionProcessor.go [hunk 9]
```diff
@@ -219,25 +220,26 @@ func (tp *TransactionProcessor) simulateTransaction(observers []*data.NodeData,
...
 			continue
 		}
 
 		// if the request was bad, return the error message
-		return nil, err
+		return nil, respCode, err
 	}
 
-	return nil, ErrSendingRequest
+	return nil, gRespCode, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:237]: same here (`all observers down/not responding != bad request `)

File: process/transactionProcessor.go [hunk 10]
```diff
@@ -257,15 +259,15 @@ func (tp *TransactionProcessor) SendMultipleTransactions(txs []*data.Transaction
...
+		return data.MultipleTransactionsResponseData{}, http.StatusBadRequest, ErrNoValidTransactionToSend
 	}
 
 	txsHashes := make(map[int]string)
 	txsByShardID := tp.groupTxsByShard(txsToSend)
 	for shardID, groupOfTxs := range txsByShardID {
 		observersInShard, err := tp.proc.GetObservers(shardID)
 		if err != nil {
-			return data.MultipleTransactionsResponseData{}, ErrMissingObserver
+			return data.MultipleTransactionsResponseData{}, http.StatusInternalServerError, ErrMissingObserver
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:270]: in other places you returned bad request

File: process/transactionProcessor.go [hunk 11]
```diff
@@ -293,21 +295,22 @@ func (tp *TransactionProcessor) SendMultipleTransactions(txs []*data.Transaction
...
 	err := tp.checkTransactionFields(tx)
 	if err != nil {
-		return "", err
+		return "", http.StatusBadRequest, err
 	}
 
 	observers, err := tp.proc.GetAllObservers()
 	if err != nil {
-		return "", err
+		return "", http.StatusInternalServerError, err
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:310]: in other places you returned bad request

File: process/transactionProcessor.go [hunk 12]
```diff
@@ -320,33 +323,34 @@ func (tp *TransactionProcessor) TransactionCostRequest(tx *data.Transaction) (st
...
 		}
 
 		// if the request was bad, return the error message
-		return "", err
+		return "", respCode, err
 
 	}
 
-	return "", ErrSendingRequest
+	return "", gRespCode, ErrSendingRequest
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:341]: `all observers down/not responding != bad request `

File: process/transactionProcessor.go [hunk 13]
```diff
@@ -383,31 +387,31 @@ func (tp *TransactionProcessor) getShardByAddress(address string) (uint32, error
...
 }
 
-func (tp *TransactionProcessor) getTxFromObservers(txHash string, reqType requestType, withResults bool) (*data.FullTransaction, error) {
+func (tp *TransactionProcessor) getTxFromObservers(txHash string, reqType requestType, withResults bool) (*data.FullTransaction, int, error) {
 	observersShardIDs := tp.proc.GetShardIDs()
 	for _, observerShardID := range observersShardIDs {
 		nodesInShard, err := tp.getNodesInShard(observerShardID, reqType)
 		if err != nil {
-			return nil, err
+			return nil, http.StatusInternalServerError, err
```
Comment (bogdan-rosianu) [inline process/transactionProcessor.go:414]: bad request of interal error?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: facade/interface.go (hunk 2 above)
Comment (andreibancioiu) [reply facade/interface.go:52]: How about even including a `status` field on the API response base?

**Review by andreibancioiu — COMMENTED** [review]
How about doing the changes in `v2`? Is it possible? (optional)

**Pull request closed without merging**